lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,218 @@
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_FILE_READ_SIZE;
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_KEY;
+import static io.trino.filesystem.tracing.Tracing.withTracing;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public final class MemoryFileSystemCache
+        implements TrinoFileSystemCache
+{
+    private final Tracer tracer;
+    private final int maxContentLengthBytes;
+    private final Cache<String, Optional<Slice>> cache;
+    private final AtomicLong largeFileSkippedCount = new AtomicLong();
+
+    @Inject
+    public MemoryFileSystemCache(Tracer tracer, MemoryFileSystemCacheConfig config)
+    {
+        this(tracer, config.getCacheTTL(), config.getMaxSize(), config.getMaxContentLength());
+    }
+
+    public MemoryFileSystemCache(Tracer tracer, Duration expireAfterWrite, DataSize maxSize, DataSize maxContentLength)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+        checkArgument(maxContentLength.compareTo(DataSize.of(1, GIGABYTE)) <= 0, "maxContentLength must be less than or equal to 1GB");
+        this.cache = EvictableCacheBuilder.newBuilder()
+                .maximumWeight(maxSize.toBytes())
+                .weigher((Weigher<String, Optional<Slice>>) (key, value) -> toIntExact(estimatedSizeOf(key) + sizeOf(value, Slice::getRetainedSize)))
+                .expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS)
+                .shareNothingWhenDisabled()
+                .recordStats()
+                .build();
+        this.maxContentLengthBytes = toIntExact(maxContentLength.toBytes());
+    }
+
+    @Override
+    public TrinoInput cacheInput(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getOrLoadFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateInput");
+            return withTracing(span, delegate::newInput);
+        }
+
+        return new MemoryInput(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public TrinoInputStream cacheStream(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getOrLoadFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateStream");
+            return withTracing(span, delegate::newStream);
+        }
+
+        return new MemoryInputStream(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public long cacheLength(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getOrLoadFromCache(key, delegate);
```

Comment:
Seems like we will load file to just to get length. Use `getIfPresent` maybe? Or maybe have a dedicated cache for lengths? Do we even need one since `TrinoInputFile` implementations cache length anyway (even `CacheInputFile`)?

Reply:
> Seems like we will load file to just to get length

I don't know of a code path where we just look at length, we always want to read contents as well after getting length.

> Do we even need one since TrinoInputFile implementations cache length anyway

That only works for within query and is tied to lifecycle of TrinoInputFile object.
We want to be able to cache length calls across queries.